### PR TITLE
Search users: API

### PR DESCRIPTION
**Pull request by mrkvon**
Backend API part for _"searching members you already know"_ -feature #530

The endpoint is `GET /api/users?search=[searchString]`, where replace `[searchString]`.

The endpoint will support search by full names, (given and family) and full usernames. The support for partial search (_search as you type_) will be dropped.

### CODE REVIEW

**Comment by mrkvon** [comment]
The basic version is finished. Should we return the logged user (self) from the search?

`GET /api/users?search=string`

The query string can use `+` instead of a space.

**Comment by simison** [comment]
Great! I'll have a look today.

>Should we return the logged user (self) from the search?

That's fine I think.

File: modules/users/server/controllers/users/users.profile.server.controller.js [hunk 1]
```diff
@@ -1051,3 +1051,51 @@ exports.validateInviteCode = function(req, res) {
...
+          { username: queryRegexp },
+          { firstName: queryRegexp },
+          { lastName: queryRegexp },
+          { displayName: queryRegexp }
+        ]
+      }
+    ] })
+    // select only the right profile properties
+    .select(exports.userMiniProfileFields + ' -_id')
+    // todo what to choose for sorting?
```
Comment (simison) [inline modules/users/server/controllers/users/users.profile.server.controller.js]: Sorting by `username` is fine. :thumbsup:

File: modules/users/tests/server/search-users.server.routes.tests.js [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+        ], done);
+      });
+    });
+
+    context('invalid request', function () {
+      it('[query string is less than 3 characters long] respond with 400', function (done) {
+        agent.get('/api/users?search=aa')
+          .expect(400)
+          .end(function (err) {
+            // @TODO test the error message
```
Comment (simison) [inline modules/users/tests/server/search-users.server.routes.tests.js]:
```
.end(function (err, res) {

    res.body.message.should.equal('...Foo...');

});

**Comment by mrkvon** [comment]
rebased with master

**Comment by mrkvon** [comment]
I was just looking into [indexes to use](https://docs.mongodb.com/manual/reference/operator/query/regex/#index-use) and noticed this: 

> Case insensitive regular expression queries generally cannot use indexes effectively. The `$regex` implementation is not collation-aware and is unable to utilize case-insensitive indexes.

So indexes won't help in this case (unless we create lowercase versions of `username`, `displayName`, `firstName`, `lastName` with indexes and perform case sensitive lowercase search :unamused:).

**Comment by simison** [comment]
This seems like worth exploring:  https://docs.mongodb.com/manual/core/index-case-insensitive/ <https://docs.mongodb.com/manual/core/index-case-insensitive/>

**Comment by mrkvon** [comment]
@simison That seems like the case-insensitive index, which can't be used by regexp search, according to the above quote. o_o

**Comment by simison** [comment]
Right; perhaps we could drop support for partial search but keep case-insensitive with that type of index.

I think MongoDB already treats stuff like `é` and `e` as same with text indices?

At this point, we could probably take the performance hit of regexp queries just fine but I'd rather not open an easy path for DoS attacks via expensive API endpoints.

**Comment by mrkvon** [comment]
Ok. Thank you @simison for the specification. I'd like to leave this to sink in for a while. When the design and (API needs) for the frontend are clear and somebody is ready to implement it, I think I can fix this relatively quickly.

For the future reference, people who will use this endpoint: The support for the _search as you type_ functionality is deprecated. It won't work in production. Only the search by full given names, family names and usernames.

The endpoint is `GET /api/users?search=[searchString]`, where replace `[searchString]`.

**Comment by simison** [comment]
@mrkvon sounds like a godo plan :thumbsup:

**Comment by Akronix** [comment]
I'm up for taking the frontend part :raising_hand_man:
There was also a designer from the Iberian collective willing to do the design part. I'll ping him.

**Comment by simison** [comment]
I'd recommend re-using HTML+CSS from contacts -tab as much as possible.

**Comment by Regenwasser** [comment]
Continuing from #530 (but not quite sure if I'm at the right place here?): 

> Agreed not to have additional UX for this and build it in the existing API.

If we can do that I'd love to add keyword search right in there, even if it takes longer for it to appear on the site. Just like adding something to the URL structure would be sweet: 

`http://trustroots.org/search?keyword=placeholder`

That would be helpful for user research purposes, without having to go into any database.

**Comment by guaka** [comment]
What's happening with this?

I don't think much design was done, but let's push this through asap - after that I'd be very happy to adapt it for users with `admin` (or rather `support`) role so that the search will be useful for this. (Using my hackety cli script for regex search on users all the time, to find someone's profile when they contact us just by email or even FB.)

**Comment by mrkvon** [comment]
Rebased. I'll re-familiarize myself with this and try to update to fit the requirements.

**Comment by mrkvon** [comment]
As of now, we're using a compound text index on `username`, `firstName` and `lastName`.

I just had a conversation with @Akronix how to improve this in a next iteration (pagination, ...). Let's follow up in an issue.

Also we may want another index (not text index) on `username`. I don't know how dots and dashes in usernames are treated now.

**Pull request merged**